chiselFrontend/src/main/scala/chisel3/core/StrongEnum.scala

```diff
@@ -0,0 +1,339 @@
+    }
+
+    if (litOption.isDefined) {
+      val index = companionObject.all.indexOf(this)
+      if (index < companionObject.all.length-1) companionObject.all(index+1)
+      else companionObject.all.head
+    } else {
+      def muxBuilder(enums: List[this.type], first_enum: this.type): this.type = enums match {
+        case e :: Nil => first_enum
+        case e :: e_next :: es => Mux(this === e, e_next, muxBuilder(e_next :: es, first_enum))
+      }
+
+      muxBuilder(companionObject.all, companionObject.all.head)
+    }
+  }
+
+  private[core] def bindToLiteral(bits: UInt): Unit = {
+    val litNum = bits.litOption.get
+    val lit = ULit(litNum, width) // We must make sure to use the enum's width, rather than the UInt's width
+    lit.bindLitArg(this)
+  }
+
+  override def bind(target: Binding, parentDirection: SpecifiedDirection): Unit = {
+    super.bind(target, parentDirection)
+
+    // If we try to annotate something that is bound to a literal, we get a FIRRTL annotation exception.
+    // To workaround that, we only annotate enums that are not bound to literals.
+    if (selfAnnotating && !litOption.isDefined) {
+      annotateEnum()
+    }
+  }
+
+  private def annotateEnum(): Unit = {
+    annotate(EnumComponentChiselAnnotation(this, enumTypeName))
+
+    if (!Builder.annotations.contains(companionObject.globalAnnotation)) {
+      annotate(companionObject.globalAnnotation)
+    }
+  }
+
+  private def enumTypeName: String = getClass.getName
+
+  // TODO: See if there is a way to catch this at compile-time
+  def checkTypeEquivalency(that: EnumType): Unit =
+    if (!typeEquivalent(that)) {
+      throw EnumTypeMismatchException(s"${this.getClass.getName} and ${that.getClass.getName} are different enum types")
+    }
+
+  def toPrintable: Printable = FullName(this) // TODO: Find a better pretty printer
+}
+
+// This is an enum type that can be connected directly to UInts. It is used as a "glue" to cast non-literal UInts
+// to enums.
+sealed private[chisel3] class UnsafeEnum(override val width: Width) extends EnumType(selfAnnotating = false) {
+  override def cloneType: this.type = getClass.getConstructor(classOf[Width]).newInstance(width).asInstanceOf[this.type]
+}
+private object UnsafeEnum extends StrongEnum[UnsafeEnum] {
+  override def checkEmptyConstructorExists(): Unit = {}
+}
+
+abstract class StrongEnum[T <: EnumType : ClassTag] {
+  private var id: BigInt = 0
+  private[core] var width: Width = 0.W
+
+  private val enum_names = getEnumNames
+  private val enum_values = mutable.ArrayBuffer.empty[BigInt]
+  private val enum_instances = mutable.ArrayBuffer.empty[T]
+
+  private def getEnumNames(implicit ct: ClassTag[T]): Seq[String] = {
+    val mirror = runtimeMirror(this.getClass.getClassLoader)
+
+    // We use Java reflection to get all the enum fields, and then we use Scala reflection to sort them in declaration
+    // order. TODO: Use only Scala reflection here
```

Comment:
Yes, there were cases where Scala reflection just pukes all over the place. I think this had to do with inner classes and fun like that, so we're still stuck with looking at id order, which is kind of hacky and nasty.
We use Scala reflection in autoclonetype, but that's fine because that's best effort - it's more or less accepted to tell the user that they need to overload cloneType (though we might not have if we were doing clean-slate design).